### PR TITLE
End-to-end planner updates

### PR DESCRIPTION
**Pull request by halil93ibrahim**
I am creating this PR instead of #286, since it includes updates in the end-to-end planner tool to include the recent [work](https://www.mdpi.com/2218-6581/11/5/109). I have also implemented the ROS1 node here, which publishes the output of the planner as a target waypoint.

### CODE REVIEW

File: src/opendr/planning/end_to_end_planning/envs/UAV_depth_planning_env.py [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+                self.ros_srv_field_set_rotation(self.robot_rotation_field, 0, goal.pose.orientation)
+                self.r.sleep()
+            except rospy.ServiceException as exc:
+                print("Service did not process request: " + str(exc))
+        else:
+            goal = PoseStamped()
+
+            goal.header.seq = 1
+            goal.header.stamp = rospy.Time.now()
+            # goal.header.frame_id = "map"
```
Comment (tsampazk) [inline src/opendr/planning/end_to_end_planning/envs/UAV_depth_planning_env.py:358]: If this is not needed feel free to remove.

**Review by tsampazk — CHANGES_REQUESTED** [review]
Thanks @halil93ibrahim for the new node and the updates! As this tool seems more involved and different in nature, i don't think we need to force the ROS node format that is followed in other packages, what do you think @ad-daniel?

Changes are good from my perspective, most comments are about minor stuff; some broken links in the docs and leftover comments.

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 2]
```diff
@@ -59,24 +66,27 @@ def download(self, path=None,
...
+                else:
+                    self.env = env
+                    self.agent = PPO("MultiInputPolicy", self.env, learning_rate=self.lr, n_steps=self.n_steps,
+                                     batch_size=self.batch_size, verbose=verbose)
             else:
                 print('env should be gym.Env')
                 return
         self.last_checkpoint_time_step = 0
-        self.mean_reward = -10
+        # self.mean_reward = -10
```
[the remark this answers is not in the document]
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:89]:
```suggestion
```

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 3]
```diff
@@ -85,8 +95,9 @@ def fit(self, env=None, logging_path='', silent=False, verbose=True):
         self.env = Monitor(self.env, filename=self.logdir)
         self.env = DummyVecEnv([lambda: self.env])
         self.agent.set_env(self.env)
-        self.agent.learn(total_timesteps=self.iters, callback=self.callback)
-        return {"last_20_episodes_mean_reward": self.mean_reward}
+        checkpoint_callback = CheckpointCallback(save_freq=1000, save_path=self.logdir, name_prefix='rl_model')
+        self.agent.learn(total_timesteps=self.iters, callback=checkpoint_callback)
+        # return {"last_20_episodes_mean_reward": self.mean_reward}
```
[the remark this answers is not in the document]
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:100]:
```suggestion
```

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 4]
```diff
@@ -159,18 +171,18 @@ def reset(self):
...
+    #     else:
+    #         return True
+    #
+    #     if x[-1] - self.last_checkpoint_time_step > self.checkpoint_after_iter:
+    #         self.last_checkpoint_time_step = x[-1]
+    #         check_point_path = Path(self.logdir,
+    #                                 'checkpoint_save' + str(x[-1]) + 'with_mean_rew' + str(self.mean_reward))
+    #         self.save(str(check_point_path))
+    #
+    #     return True
```
[the remark this answers is not in the document]
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:188]:
```suggestion
```

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 5]
```diff
@@ -22,6 +22,7 @@
 from stable_baselines3.common.monitor import Monitor
```
[the remark this answers is not in the document]
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:22]: I remember it has a need, but the test code passes without importing rospy. I will remove it.

File: src/opendr/planning/end_to_end_planning/envs/UAV_depth_planning_env.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+                resp2 = self.ros_srv_get_field(self.robot_node_id, 'rotation', False)
+            except rospy.ServiceException as exc:
+                print("Service did not process request: " + str(exc))
+            self.robot_translation_field = resp1.field
+            self.robot_rotation_field = resp2.field
+
+    def step(self, action):
+        # if self.current_position == PoseStamped().pose.position:
+        #     rospy.loginfo("Gym environment is not reading mavros position")
+        #     return self.observation_space.sample(), np.random.random(1), False, {}
```
[the remark this answers is not in the document]
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/envs/UAV_depth_planning_env.py]: Done!

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 7]
```diff
@@ -21,31 +21,37 @@
...
 from opendr.engine.learners import LearnerRL
 from opendr.engine.constants import OPENDR_SERVER_URL
 
-__all__ = ["rospy", ]
+# __all__ = ["rospy", ]
 
 
 class EndToEndPlanningRLLearner(LearnerRL):
-    def __init__(self, env, lr=3e-4, n_steps=1024, iters=int(5e4), batch_size=64, checkpoint_after_iter=500,
+    def __init__(self, env=None, lr=3e-4, n_steps=1024, iters=int(1e5), batch_size=64, checkpoint_after_iter=500,
```
Comment (ad-daniel) [inline src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:33]: Changes to default values should be updated accordingly in the documentation file as well (`end-to-end-planning.md`)

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 8]
```diff
@@ -21,31 +21,37 @@
...
 from stable_baselines3.common.monitor import Monitor
 from stable_baselines3.common.vec_env import DummyVecEnv
-from stable_baselines3.common.results_plotter import load_results, ts2xy
+from stable_baselines3.common.callbacks import CheckpointCallback
 
 from opendr.engine.learners import LearnerRL
 from opendr.engine.constants import OPENDR_SERVER_URL
 
-__all__ = ["rospy", ]
+# __all__ = ["rospy", ]
```
Comment (ad-daniel) [inline src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:29]: if indeed unnecessary, then I think we can remove it altogether not just comment it (likewise in line 15)

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 9]
```diff
@@ -59,24 +65,26 @@ def download(self, path=None,
             urlretrieve(url=url, filename=file_destination)
         return file_destination
 
-    def fit(self, env=None, logging_path='', silent=False, verbose=True):
+    def fit(self, env=None, logging_path='', verbose=True):
```
Comment (ad-daniel) [inline src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:68]: same, documentation needs to be updated

**Review by ad-daniel — CHANGES_REQUESTED** [review]
Yes, I think this tool is a bit of an exception and it doesn't make much sense to enforce the same constraints as the others

File: docs/reference/end-to-end-planning.md [hunk 10]
```diff
@@ -103,7 +103,10 @@ Parameters:
```
Comment (tsampazk) [inline docs/reference/end-to-end-planning.md:103]: There's a link to the learner on line 12 that's broken in a similar fashion as the others that i missed earlier, need to add `../../`.

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py (hunk 8 above)
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py]: removed

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py (hunk 9 above)
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:65]: updated

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py (hunk 7 above)
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:30]: updated

File: docs/reference/end-to-end-planning.md (hunk 10 above)
Comment (halil93ibrahim) [reply docs/reference/end-to-end-planning.md:101]: added

**Comment by halil93ibrahim** [comment]
Hello, I have implemented your comments. Please let me know If there is further.
For tests to run successfully, preferably roscore and the Webots world provided are expected to run. What should I do to perform those?

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py [hunk 11]
```diff
@@ -85,8 +90,8 @@ def fit(self, env=None, logging_path='', silent=False, verbose=True):
         self.env = Monitor(self.env, filename=self.logdir)
         self.env = DummyVecEnv([lambda: self.env])
         self.agent.set_env(self.env)
-        self.agent.learn(total_timesteps=self.iters, callback=self.callback)
-        return {"last_20_episodes_mean_reward": self.mean_reward}
+        checkpoint_callback = CheckpointCallback(save_freq=1000, save_path=self.logdir, name_prefix='rl_model')
+        self.agent.learn(total_timesteps=self.iters, callback=checkpoint_callback)
 
     def eval(self, env):
         """
```
Comment (tsampazk) [inline src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:97]: Below on line 107 `# env = Monitor(env, filename=self.logdir)` could be removed if not needed.

**Review by tsampazk — CHANGES_REQUESTED** [review]
Thanks @halil93ibrahim! Two very minor comments and it's good to go from my perspective. 

> For tests to run successfully, preferably roscore and the Webots world provided are expected to run. What should I do to perform those?

I'm not exactly sure how to handle this.

File: src/opendr/planning/end_to_end_planning/e2e_planning_learner.py (hunk 11 above)
Comment (halil93ibrahim) [reply src/opendr/planning/end_to_end_planning/e2e_planning_learner.py:97]: removed!

**Review by tsampazk — DISMISSED** [review]
Thank you!

**Review by ad-daniel — APPROVED** [review]
Thank you

> For tests to run successfully, preferably roscore and the Webots world provided are expected to run. What should I do to perform those?

You mean for the integrated tests (CI) ? The CI machines don't have GPU and running simulations CPU-only will likely result in very long tests, and the simulation also relied on a older version of webots (~1.5GB at the time) and required Ardupilot. As such we decided not to go that route and instead have a test that just ensures the tool is functional (i.e. can be installed without dependency conflicts and tests some functionality). An easier approach could be to record locally the simulation data (sensor data) and use that in the test without needing to install webots itself. At the time we opted for the current test, do you think a different approach would be better now?

**Review by tsampazk — APPROVED** [review]

**Comment by halil93ibrahim** [comment]
> You mean for the integrated tests (CI) ? The CI machines don't have GPU and running simulations CPU-only will likely result in very long tests, and the simulation also relied on a older version of webots (~1.5GB at the time) and required Ardupilot. As such we decided not to go that route and instead have a test that just ensures the tool is functional (i.e. can be installed without dependency conflicts and tests some functionality). An easier approach could be to record locally the simulation data (sensor data) and use that in the test without needing to install webots itself. At the time we opted for the current test, do you think a different approach would be better now?

I meant the CI tests. Now, the tests are passing similar to the one we did before. The tool does not require Ardupilot in this version, but still needs the old version of Webots. So, I could think of a different approach when I complete the implementations in Webots 2022b.
I believe the PR is now ready to merge. Please let me know if a change is needed.

**Review by ad-daniel — APPROVED** [review]
Ah I see, having webots on the loop would be ideal, but like I mentioned it will depend on the performance. If the tests take too long is not great and without GPU that might be the case. We can explore the option again when the version for R2022b is ready.
Looks good to me as well, thank you!

**Pull request merged**